components/CollapsibleTable.tsx

```diff
@@ -0,0 +1,157 @@
+
+import styles from "./CollapsibleTable.module.css";
+import Tabs from "./Tabs";
+
+import Row from "./Row";
+
+import {
+  ID,
+  DependencyData,
+  DependencyMap,
+  DependencyListSingleDep,
+  findRank,
+  rankToDepColour,
+  semVerToString,
+  SemVer,
+} from "./dataProcessing";
+import { Circle } from "@mui/icons-material";
+
+// Creates the whole table based on sample data
+const CollapsibleTable = (rows: JSX.Element[]) => {
+  return (
+    <div>
+      <TableContainer component={Paper}>
+        <Table size="small" aria-label="collapsible table">
+          <colgroup>
+            <col style={{ width: "0%", backgroundColor: "#f6f6f6" }} />
+            <col style={{ width: "0%" }} />
+            <col style={{ width: "75%" }} />
+            <col style={{ width: "25%" }} />
+            <col style={{ width: "0%" }} />
+          </colgroup>
+          <TableHead>
+            <TableRow>
+              <TableCell></TableCell>
+              <TableCell></TableCell>
+              <TableCell>Name</TableCell>
+              <TableCell>Version</TableCell>
+              <TableCell>Link</TableCell>
+            </TableRow>
+          </TableHead>
+          <TableBody>{rows}</TableBody>
+        </Table>
+      </TableContainer>
+    </div>
+  );
+};
+
+function makeSubRow(
+  data: DependencyListSingleDep,
+  rank: number,
+  dependencyMap: DependencyMap
+) {
+  const dependencyData = dependencyMap.get(data.id);
+  const str =
+    dependencyData.name +
+    ": " +
+    semVerToString(data.version) +
+    " -> " +
+    semVerToString(dependencyData.version);
+  const [colour, borderColour, colourIndex] = rankToDepColour(rank);
+  const dep = (
+    <TableRow>
+      <col style={{ width: "0%" }} />
+      <col style={{ width: "75%" }} />
+      <TableCell>
+        <Circle style={{ color: colour }} />
+      </TableCell>
+      <TableCell>{str}</TableCell>
+    </TableRow>
+  );
+  return dep;
+}
+
+function makeInverseSubRow(
+  data: DependencyListSingleDep,
+  rank: number,
+  dependencyMap: DependencyMap
+) {
+  const dependencyData = dependencyMap.get(data.id);
+  const str = dependencyData.name + ": " + semVerToString(data.version);
+  const [colour, borderColour, colourIndex] = rankToDepColour(rank);
+  const dep = (
+    <TableRow>
+      <col style={{ width: "0%" }} />
+      <col style={{ width: "75%" }} />
+      <TableCell>
+        <Circle style={{ color: colour }} />
+      </TableCell>
+      <TableCell>{str}</TableCell>
+    </TableRow>
+  );
+  return dep;
+}
+
+const makeCollapsibleTable = (JSObject: DependencyData) => {
+  //TODO: Rather than taking in the raw objet, take a structure that can be directly converted into JSX
+  let rowList: JSX.Element[] = [];
+
+  for (const dep of JSObject.deps) {
+    const data = JSObject.depMap.get(dep.id);
+    let internalSubRows: JSX.Element[] = [];
+    let externalSubRows: JSX.Element[] = [];
+    let userSubRows: JSX.Element[] = [];
+
+    let minRank = 2;
+
```

Comment:
a short description of what this loop is / what ranks are might improve maintainability